SWITCHING_EXTRUDER improvement - double switching extruder support

fixoid opened the pull request:
A small change to the  `SWITCHING_EXTRUDER `feature that adds second switching extruder support.

fixoid committed:
double switching extruder support

fixoid commented:
Tested with this configuration:
```
#define EXTRUDERS 4 

#define SINGLENOZZLE

#define SWITCHING_EXTRUDER
#if ENABLED(SWITCHING_EXTRUDER)  
  #define SWITCHING_EXTRUDER_SERVO_NR 0
  #define SWITCHING_EXTRUDER_SERVO_ANGLES { 0, 90, 0, 90 } // Angles for E0, E1[, E2, E3]
  #if (EXTRUDERS == 4) 
    #define SWITCHING_EXTRUDER_E23_SERVO_NR 1
  #endif 
#endif
```

Setting for the case described in #6260 - 4 filaments with 2 steppers + singlenozzle, but for new style `SWITCHING_EXTRUDER `feature.

Works well (at first sight).

fixoid commented:
Some code optimization in `Marlin_main` needed, will do in next 5-6 hours

thinkyhead commented:
Ok, let me know when it's ready. Do you know how to squash commits? I'd like to reduce this down to one. If you don't know how to do that, no problem. I can do it as part of the merge.

fixoid commented:
Ok, i squashed all commits exept last, cause i want to give some coments and ask few questions about it.

1. I added EXTRUDERS check for `HOTENDS <= 1 ` (str 9817) to avoid servo movemets when trying to turn on invalid tool. But maybe it was some special reason why you not iclude it before, then i will move it down in to  ` #if ENABLED(SWITCHING_EXTRUDER)` block.

2. In case of `HOTENDS <= 1` 
first: `active_extruder = tmp_extruder;` (str. 9821), 
then  `move_extruder_servo(active_extruder);` (str. 9821) 
and it works right.
But `#if HOTENDS > 1` (str. 9531) 
there are `move_extruder_servo(active_extruder);` (str.9671) 
before `active_extruder = tmp_extruder;` (str. 9776) 
and it works inverted.

I think str. 9671 must be changed to `move_extruder_servo(tmp_extruder);`

thinkyhead commented:
All good. I've sent you a PR with the patches.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -347,8 +347,8 @@
     #error "SWITCHING_EXTRUDER and DUAL_X_CARRIAGE are incompatible."
-  #elif EXTRUDERS != 2
-    #error "SWITCHING_EXTRUDER requires exactly 2 EXTRUDERS."
+  #elif (EXTRUDERS != 2) && (EXTRUDERS != 4)
```
😁 
```cpp
#elif EXTRUDERS & 1
  #error "SWITCHING_EXTRUDER requires an even number of EXTRUDERS."
. . .
```

fixoid commented:
> I've sent you a PR with the patches.

Sorry, not sure that I understand you right, i havent recieved anything from you

In next 7-8 hours i want to make one more commit and squash all together.

thinkyhead commented:
fixoid/Marlin#1

fixoid commented:
Got it, will check in few hours

thinkyhead committed:
Suggested patches (#1)

* Suggested patches

fixoid committed:
extruder indirection for switching extruder fix and misc

fixoid committed:
Update mfqp

fixoid committed:
Update Marlin_main.cpp

fixoid committed:
Update Marlin_main.cpp

fixoid committed:
Update Configuration.h

fixoid committed:
Update Conditionals_LCD.h

fixoid committed:
1

fixoid committed:
Merge remote-tracking branch 'origin/patch-3' into patch-3

# Conflicts:
#	Marlin/Conditionals_LCD.h
#	Marlin/Configuration.h
#	Marlin/Marlin_main.cpp

fixoid committed:
Update Conditionals_LCD.h

fixoid committed:
Update Configuration.h

fixoid committed:
Update Marlin_main.cpp

fixoid committed:
Update mfqp

fixoid committed:
Update Conditionals_LCD.h

fixoid committed:
Update pins.h

fixoid commented:
@thinkyhead , thanks for patches!
Already merged to this PR. Also i made few fixes after it to make it work properly.

Tested with 2,3 and 4 `EXTRUDERS` - works well.
Can't test with `#define EXTRUDERS 5` because of RAMPS 

P.S. Sorry for the huge amount of commits. I did something wrong with squashing and had a version conflict.

fixoid commented on Marlin/Configuration.h:
```diff
@@ -142,7 +142,10 @@
   #define SWITCHING_EXTRUDER_SERVO_NR 0
-  #define SWITCHING_EXTRUDER_SERVO_ANGLES { 0, 90 } // Angles for E0, E1
+  #define SWITCHING_EXTRUDER_SERVO_ANGLES { 0, 90 } // Angles for E0, E1[, E2, E3]
+  #if EXTRUDERS > 3
```
Condition changed to avoid unnecessary definition with 3 extruders

thinkyhead replied:
```cpp
// Angles for E0, E1[, E2[, E3[, E4]]]
```

fixoid commented on Marlin/Marlin_main.cpp:
```diff
@@ -9483,25 +9483,39 @@ inline void gcode_M999() {
+    constexpr int16_t angles[] = SWITCHING_EXTRUDER_SERVO_ANGLES;
+    static_assert(COUNT(angles) / 2 >= E_STEPPERS, "SWITCHING_EXTRUDER_SERVO_ANGLES needs 2 angles per servo.");
+
+    #if EXTRUDERS & 1
```
changed E_STEPPERS on EXTRUDERS
with E_STEPPERS, if EXTRUDERS == 2 then if (e < EXTRUDERS - 1) won't switch servo in case of e=1

fixoid replied:
also there is no way to check 3/5 EXTRUDERS  option with E_STEPPERS

thinkyhead replied:
I'm pretty sure `E_STEPPERS` will always be equal to `EXTRUDERS` with this type of setup, and this feature relates to the steppers, not the nozzles. `EXTRUDERS` should only very rarely be used, since it represents neither nozzles nor steppers, but rather logical extruding entities.

fixoid replied:
@thinkyhead , maybe I'm wrong, but
with `SWITCHING_EXTRUDER `we have only one stepper per two tools.
And in this void `e `equal to tool 0,1,3,4 or 5.
with 2 `EXTRUDERS `we will have only one `E_STEPPERS`.
with 3 or 4 `EXTRUDERS `- two `E_STEPPERS`.

So we can't use `E_STEPPERS ` with  `e` here.

fixoid replied:
@thinkyhead ,
Then User is configuring Marlin, he can operate only with definition of `EXTRUDERS `in **Configuration.h**
With `SWITCHNG_EXTRUDER` we have
`EXTRUDERS `= x (user defined)
`HOTENDS `= `EXTRUDERS`
`E_STEPPERS `= `EXTRUDERS /2` (for 2 or 4 extruders) or `(EXTRUDERS + 1) / 2` (for 3 or 5 extruders)

So we’ll have the same `E_STEPPERS `for 3 and 4 `EXTRUDERS`, 
if `EXTRUDERS `= 3, we do not need to perform any switching of servo, when activating 3-rd tool (`T2`), because we have one Switching extruder(1-st Stepper for `T0 `and `T1`) and one classic extruder (2-nd Stepper for `T2`)
If we try to operate with `E_STEPPERS `here, Marlin will not know if it necessary to switch servo, when we activate `T2`, because it can’t recognize how many tools (`EXTRUDERS`) are defined – 3 or 4

fixoid commented on Marlin/Marlin_main.cpp:
```diff
@@ -9483,25 +9483,39 @@ inline void gcode_M999() {
+      if (e < EXTRUDERS - 1)
+    #endif
+    {
+      #if EXTRUDERS > 3
```
reorganised to avoid attemps to switch 2-nd servo  in case of `#def EXTRUDERS 3`

fixoid commented on Marlin/Marlin_main.cpp:
```diff
@@ -9521,10 +9535,10 @@ void tool_change(const uint8_t tmp_extruder, const float fr_mm_s/*=0.0*/, bool n
   #else // !MIXING_EXTRUDER || MIXING_VIRTUAL_TOOLS <= 1
 
-    #if HOTENDS > 1
+    if (tmp_extruder >= EXTRUDERS)
```
Some code clearance

fixoid commented:
@thinkyhead - I would be grateful if you could make a review and give some comments and suggestions, if it needed. 
My vacation is approaching and I would like to finish this before it :)
Thanks.

fixoid commented:
@thinkyhead , I would be happy if you find time to make a review and give comments so that I can correct the errors before Saturday.
Sorry for annoing!

thinkyhead commented:
> Sorry for the huge amount of commits. I did something wrong with squashing and had a version conflict.

That's when `git diff` can be used to skip all the merge complications and apply all the differences between one branch and another all at once.
```bash
git fetch MarlinFirmware                                    ;# Get the latest code

git checkout my_messy_branch                                ;# Make sure messy branch fits with bugfix
git rebase MarlinFirmware/bugfix-1.1.x                      ;# using rebase, moving commits to the end.

git checkout MarlinFirmware/bugfix-1.1.x -b my_temp_branch  ;# Make a new copy of the latest bugfix

git diff HEAD my_messy_branch | git apply                   ;# Apply changes from messy branch (assuming it's up to date)

git add *                                                   ;# This does one commit, but you can use Github Desktop to
git commit -m "All changes as one commit."                  ;# cherry-pick changes to keep, make a few commits, whatever...

git checkout my_messy_branch                                ;# Now update the messy branch... and the PR!
git reset --hard my_temp_branch                             ;# Just point it at our nice new commits
git push -f                                                 ;# Update the PR

git branch -D my_temp_branch                                ;# Done with this
```

fixoid commented on Marlin/Marlin_main.cpp:
```diff
@@ -9483,25 +9483,39 @@ inline void gcode_M999() {
+  inline void move_extruder_servo(const uint8_t e) {
+    stepper.synchronize();
+    constexpr int16_t angles[] = SWITCHING_EXTRUDER_SERVO_ANGLES;
+    static_assert(COUNT(angles) / 2 >= E_STEPPERS, "SWITCHING_EXTRUDER_SERVO_ANGLES needs 2 angles per servo.");
```
@thinkyhead, this test:
`static_assert(COUNT(angles) / 2 >= E_STEPPERS, "SWITCHING_EXTRUDER_SERVO_ANGLES needs 2 angles per servo.");`
is not valid for `EXTRUDERS 3`
because in this case - 
`E_STEPPERS` = (3 + 1) / 2 = 2
and 
`#define SWITCHING_EXTRUDER_SERVO_ANGLES { 0, 90 } `//we need only two angles - for T0 and T1
So:
2 / 2 < 2 and:
```
exit status 1
static assertion failed: SWITCHING_EXTRUDER_SERVO_ANGLES needs 2 angles per servo.
```
the same for `EXTURUDERS 5`
`E_STEPPERS` = (5 + 1) / 2 = 3
`#define SWITCHING_EXTRUDER_SERVO_ANGLES { 0, 90, 0, 90 }` //4 angles - for T0/T1, T2/T3
So:
4/2 < 3
____
`static_assert(COUNT(angles) == 2 || COUNT(angles) == 4, "SWITCHING_EXTRUDER_SERVO_ANGLES needs 2 angles per servo.");` 
will work here

fixoid committed:
Merge branch 'bugfix-1.1.x' into patch-3

fixoid committed:
Create Marlin_main.cpp

fixoid commented:
@thinkyhead ,
> That's when git diff can be used to skip all the merge complications and apply all the differences between one branch and another all at once.

I tried to do it several times, but something went wrong and I could not make correct single commit


> I can do it as part of the merge.

I think it's good idea for this PR,
or I can make new branch, repeat all changes in it  and make another PR to awoid conflicts

fixoid committed:
Update Marlin_main.cpp

thinkyhead commented:
Re-built this as #7096 using `git diff`.
